File: src/data/courses/dsa/dynamic-programming/classic-dp/01-knapsack/code/python/01_knapsack.py
class Solution:
    def knapsack(self, W, wt, val):
        dp = [0] * (W + 1)
        for i in range(len(wt)):
            for w in range(W, wt[i]-1, -1):  # Right to left!
                dp[w] = max(dp[w], dp[w-wt[i]] + val[i])
        return dp[W]

File: src/data/courses/dsa/dynamic-programming/classic-dp/01-knapsack/code/cpp/01_knapsack.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    int knapsack(int W, vector<int>& wt, vector<int>& val) {
        int n = wt.size();
        vector<int> dp(W + 1, 0);
        for (int i = 0; i < n; i++)
            for (int w = W; w >= wt[i]; w--) // RIGHT to LEFT — prevents reuse
                dp[w] = max(dp[w], dp[w - wt[i]] + val[i]);
        return dp[W];
    }
};
